### PR TITLE
Added a way to provide the database files to the DatabasePeerManager module.

### PR DESCRIPTION
**Pull request by jferlisi**
Created an interface to provide database files at the time of a peer connection. This allows applications that create databases in directories other than the default one to have their data viewable in the Database section.

### CODE REVIEW

File: stetho/src/main/java/com/facebook/stetho/inspector/database/DatabasePeerManager.java [hunk 1]
```diff
@@ -30,18 +29,43 @@
...
+   * Constructs the object with a {@link DatabaseFilesProvider} that supplies the database files
+   * from {@link Context#databaseList()}.
+   *
+   * @param context the context
+   */
+  public DatabasePeerManager(final Context context) {
+    this(context, new DatabaseFilesProvider() {
+      @Override public Set<File> getDatabaseFiles() {
+        Set<File> databaseFiles = new HashSet<File>();
+        for (String filename : context.databaseList())
```
Comment (jasta) [inline stetho/src/main/java/com/facebook/stetho/inspector/database/DatabasePeerManager.java]: touché, I didn't know about this API :)

File: stetho/src/main/java/com/facebook/stetho/inspector/database/DatabasePeerManager.java [hunk 2]
```diff
@@ -30,18 +29,43 @@
...
+  /**
+   * Constructs the object with a {@link DatabaseFilesProvider} that supplies the database files
+   * from {@link Context#databaseList()}.
+   *
+   * @param context the context
+   */
+  public DatabasePeerManager(final Context context) {
+    this(context, new DatabaseFilesProvider() {
+      @Override public Set<File> getDatabaseFiles() {
+        Set<File> databaseFiles = new HashSet<File>();
```
Comment (jasta) [inline stetho/src/main/java/com/facebook/stetho/inspector/database/DatabasePeerManager.java]: Prefer a list.  `HashSet` has ordering and performance implications that are seemingly unnecessary here.  Note that Android uses a list itself assuredly for this same reason.

File: stetho/src/main/java/com/facebook/stetho/inspector/database/DatabasePeerManager.java [hunk 3]
```diff
@@ -30,18 +29,43 @@
...
 
-  public DatabasePeerManager(Context context) {
+  /**
+   * Constructs the object with a {@link DatabaseFilesProvider} that supplies the database files
+   * from {@link Context#databaseList()}.
+   *
+   * @param context the context
+   */
+  public DatabasePeerManager(final Context context) {
+    this(context, new DatabaseFilesProvider() {
```
Comment (jasta) [inline stetho/src/main/java/com/facebook/stetho/inspector/database/DatabasePeerManager.java]: I'm not a fan of having the default applied internally.  Instead, create a `DefaultDatabaseFilesProvider` class and modify all constructors to require an instance of `DatabaseFilesProvider`.  Then, the default will be to merely invoke:
```java
new Database(context, new DefaultDatabaseFilesProvider(context));
```

File: stetho/src/main/java/com/facebook/stetho/inspector/database/DatabasePeerManager.java (hunk 1 above)
Comment (jasta) [inline stetho/src/main/java/com/facebook/stetho/inspector/database/DatabasePeerManager.java]: Oh, so I did.  Well that's weird, I have no memory of this at all.  *facepalm*

File: stetho/src/main/java/com/facebook/stetho/inspector/database/DatabasePeerManager.java [hunk 4]
```diff
@@ -30,18 +29,43 @@
...
+      }
+    });
+  }
+
+  /**
+   * @param context the context
+   * @param mDatabaseFilesProvider a database file name provider, can be null
+   */
+  public DatabasePeerManager(Context context,
+      DatabaseFilesProvider mDatabaseFilesProvider) {
```
Comment (jasta) [inline stetho/src/main/java/com/facebook/stetho/inspector/database/DatabasePeerManager.java]: style: `databaseFilesProvider`

File: stetho/src/test/java/com/facebook/stetho/inspector/database/DatabasePeerManagerTest.java [hunk 5]
```diff
@@ -2,28 +2,35 @@
...
-     };
-     String[] expected = {
-         "foo.db",
-         "bar.db",
-         "baz.db", "baz.db-somethingelse",
-         "dangling.db-journal",
-     };
-     List<String> tidied = DatabasePeerManager.tidyDatabaseList(databases);
-     assertArrayEquals(expected, tidied.toArray());
+    Set<File> databases = new HashSet<File>();
```
Comment (jasta) [inline stetho/src/test/java/com/facebook/stetho/inspector/database/DatabasePeerManagerTest.java]: Prefer not having to modify this test.  I actually wanted these APIs to preserve order.

**Comment by jasta** [comment]
I'm curious for your thoughts at a high level on the API for configuring modules.  I believe at present you are writing something like:
```java
new InspectorModulesProvider() {
  public Iterable<> get() {
    HashSet<> modules = new HashSet<>();
    Collections.add(modules, Stetho.defaultInspectorModulesProvider(context).get());
    replaceByType(Database.class, new Database(context, new MyDatabaseProvider(...)));
    return modules;
  }
}
```

We're doing this internally in FB4A/Messenger, but I'm really unhappy with it.  Seems like these use cases should be expressed more conveniently, perhaps with some kind of pattern like:
```java
Stetho.newInspectorModulesBuilder(context)
   .addDefaults()
   .add(new Database(...));
   .build();
```

Internally `addDefaults` would be a boolean which operated like:
```java
for (ChromeDevtoolsDomain domain : mCustomModules) {
  modules.add(domain);
  customModuleNames.add(InspectorUtil.getDomainName(domain));
}
if (mAddDefaults) {
  if (!customModuleNames.contains("Database")) {
    modules.add(...);
  }
  ...
}
```

File: stetho/src/main/java/com/facebook/stetho/inspector/database/DatabasePeerManager.java (hunk 2 above)
Comment (jferlisi) [inline stetho/src/main/java/com/facebook/stetho/inspector/database/DatabasePeerManager.java]: Will do.

File: stetho/src/main/java/com/facebook/stetho/inspector/database/DatabasePeerManager.java (hunk 3 above)
Comment (jferlisi) [inline stetho/src/main/java/com/facebook/stetho/inspector/database/DatabasePeerManager.java]: I'm not typically a fan of hiding the default logic that way, but my sense of preserving the API won the battle.

I'll create a default implementation and deprecate existing constructor.

File: stetho/src/main/java/com/facebook/stetho/inspector/database/DatabasePeerManager.java (hunk 4 above)
Comment (jferlisi) [inline stetho/src/main/java/com/facebook/stetho/inspector/database/DatabasePeerManager.java]: I'll change the name.

Do you mean move the parameter `Context context` to it's own line?

File: stetho/src/test/java/com/facebook/stetho/inspector/database/DatabasePeerManagerTest.java (hunk 5 above)
Comment (jferlisi) [inline stetho/src/test/java/com/facebook/stetho/inspector/database/DatabasePeerManagerTest.java]: Changing it to a `List` should mean that the test can pretty much go back to what it was. Are you OK with the parameter on `tidyDatabaseList` changing from an array to a list?

File: stetho/src/main/java/com/facebook/stetho/inspector/database/DatabasePeerManager.java (hunk 4 above)
Comment (jasta) [inline stetho/src/main/java/com/facebook/stetho/inspector/database/DatabasePeerManager.java]: Yes.  Our style is to put each parameter, even the first, on its own line once we have to wrap.

File: stetho/src/test/java/com/facebook/stetho/inspector/database/DatabasePeerManagerTest.java (hunk 5 above)
Comment (jasta) [inline stetho/src/test/java/com/facebook/stetho/inspector/database/DatabasePeerManagerTest.java]: Yeah that change is fine I think.

**Comment by jferlisi** [comment]
> I'm curious for your thoughts at a high level on the API for configuring modules.

I'm still playing around with things so I haven't given it a lot of thought. Right now what I have done is copied the list of modules from the defaults and add each of them individually.

```java
        ArrayList<ChromeDevtoolsDomain> modules = new ArrayList<ChromeDevtoolsDomain>();
        modules.add(new Console());
        modules.add(new CSS());
        modules.add(new Debugger());
        modules.add(new DOM());
        modules.add(new DOMStorage(context));
        modules.add(new HeapProfiler());
        modules.add(new Inspector());
        modules.add(new Network(context));
        modules.add(new Page(context));
        modules.add(new Profiler());
        modules.add(new Runtime());
        modules.add(new Worker());
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB) {
          modules.add(new Database(context, myDatabaseFilesProvider));
        }
        return modules;
```

I do like your proposed method though.

File: stetho/src/main/java/com/facebook/stetho/inspector/database/DatabasePeerManager.java [hunk 6]
```diff
@@ -30,18 +30,37 @@
...
-  public DatabasePeerManager(Context context) {
+  /**
+   * Constructs the object with a {@link DatabaseFilesProvider} that supplies the database files
+   * from {@link Context#databaseList()}.
+   *
+   * @param context the context
+   * @deprecated use
+   */
+  @Deprecated
+  public DatabasePeerManager(final Context context) {
```
Comment (jasta) [inline stetho/src/main/java/com/facebook/stetho/inspector/database/DatabasePeerManager.java]: nit: don't need final

File: stetho/src/main/java/com/facebook/stetho/inspector/database/DatabasePeerManager.java [hunk 7]
```diff
@@ -30,18 +30,37 @@
...
+    this(context, new DefaultDatabaseFilesProvider(context));
+  }
+
+  /**
+   * @param context the context
+   * @param databaseFilesProvider a database file name provider, can be null
+   */
+  public DatabasePeerManager(Context context, DatabaseFilesProvider databaseFilesProvider) {
```
Comment (jasta) [inline stetho/src/main/java/com/facebook/stetho/inspector/database/DatabasePeerManager.java]: style: new line; mark `databaseFilesProvider` with `@Nullable`

File: stetho/src/main/java/com/facebook/stetho/inspector/database/DatabasePeerManager.java [hunk 8]
```diff
@@ -30,18 +30,37 @@
...
     mContext = context;
+    this.mDatabaseFilesProvider = databaseFilesProvider;
```
Comment (jasta) [inline stetho/src/main/java/com/facebook/stetho/inspector/database/DatabasePeerManager.java]: nit: drop `this`

File: stetho/src/main/java/com/facebook/stetho/inspector/database/DatabasePeerManager.java (hunk 7 above)
Comment (jasta) [inline stetho/src/main/java/com/facebook/stetho/inspector/database/DatabasePeerManager.java]: Also, mark `mDatabaseFilesProvider` with `@Nullable` and then make things not crash if it's null.  Or drop this null support :)

**Comment by jasta** [comment]
Looks good, clean up the minor nits then rebase/squash and we can merge.

File: stetho/src/main/java/com/facebook/stetho/inspector/database/DatabasePeerManager.java (hunk 6 above)
Comment (jferlisi) [inline stetho/src/main/java/com/facebook/stetho/inspector/database/DatabasePeerManager.java]: Certainly not anymore. I'll remove it.

File: stetho/src/main/java/com/facebook/stetho/inspector/database/DatabasePeerManager.java (hunk 7 above)
Comment (jferlisi) [inline stetho/src/main/java/com/facebook/stetho/inspector/database/DatabasePeerManager.java]: I'll drop the null support. Originally I had the `DatabaseFilesProvider` was optional and added to the `context.databaseList()`. I forgot to remove the comments once that interface became the only thing providing the file names.

Comment (jferlisi) [inline stetho/src/main/java/com/facebook/stetho/inspector/database/DatabasePeerManager.java]: > style: new line

I'm not sure where you want to the new line.

Comment (jasta) [inline stetho/src/main/java/com/facebook/stetho/inspector/database/DatabasePeerManager.java]: Oops, misread :)

**Comment by jferlisi** [comment]
Do you happen to have an Intellij code style file published anywhere?

**Comment by jasta** [comment]
@jferlisi no, working on it tho :)

File: stetho/src/main/java/com/facebook/stetho/inspector/database/DatabasePeerManager.java (hunk 8 above)
Comment (jferlisi) [inline stetho/src/main/java/com/facebook/stetho/inspector/database/DatabasePeerManager.java]: It is gone.

**Comment by jferlisi** [comment]
@jasta Pushed changes.

**Comment by jasta** [comment]
Looks good, please squash into 1 commit (http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html), then `git push -f` into your github fork branch.

**Comment by jferlisi** [comment]
@jasta Done, thanks for the help on this.

**Pull request merged**